### PR TITLE
Feature: Add react-navigation support

### PR DESCRIPTION
**Pull request by Drakeoon**
### Tasks:
- [x] Add node task for development
- [x] Add option to select `react-navigation` when initializing new project
- [x] Add setup & development notes in README.md
- [x] Fix husky error when initializing new project by running `git init` in new folder
- [x] Add ability to choose react-navigation variant during new project initialization
- [x] Implement Bottom Tab navigation setup
- [x] Improve component reexports - `import Navigation from '~/navigation/index'` vs `import Navigation from '~/navigation'`
- [x] Add remaining Types if the flow is approved (`object` => `type`)
- [x] Add `native-stack` navigation

Side note: @dsznajder you may experience the issue with white background flickering when changing routes, but if you configure your device's theme under Developer Settings to use Dark Mode, it will get fixed. I think we shouldn't use black background when user has Light Mode enabled, what do you think?

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -35,6 +35,8 @@
     "registry": "https://registry.npmjs.org/"
   },
   "scripts": {
+    "dev": "yarn prepare && yarn build DevProject",
+    "build": "./bin/create-react-native-project",
```
Comment (dsznajder) [inline package.json:39]: This should probably run `yarn prepare` before.

**Comment by dsznajder** [comment]
@Drakeoon True. We can change it to white as default, and in future, add light/dark mode support and integration.

File: src/packages.ts [hunk 2]
```diff
@@ -62,6 +64,10 @@ const getPackagesToInstall = (options: Options) => {
     },
   );
 
+  if (options.modules.navigation) {
+    options.modules.gestureHandler = true;
+  }
```
Comment (dsznajder) [inline src/packages.ts:69]: Shouldn't be placed together with react-navigation as `module`?

File: templates/common/App.tsx [hunk 3]
```diff
@@ -12,6 +8,16 @@ import GraphQLClient from '~/services/GraphQL';
...
 <% } %>
+
+<% if (modules.navigation) { %>
+import Navigation from '~/navigation/index'
+<% } else { %>
+import Block from '~/components/Block';
+import Typography from '~/components/Typography';
+import palette from '~/styles/palette';
+
+import { StatusBar, StyleSheet } from 'react-native';
```
Comment (dsznajder) [inline templates/common/App.tsx:19]: Modules should be placed on top of file, before relative imports

File: templates/common/ios/Podfile [hunk 4]
```diff
@@ -1,5 +1,3 @@
-source 'https://github.com/CocoaPods/Specs.git'
```
Comment (dsznajder) [inline templates/common/ios/Podfile:1]: 😅

File: templates/common/ios/{%- project.name %}/AppDelegate.m [hunk 5]
```diff
@@ -51,7 +51,7 @@ - (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(
   if (@available(iOS 13.0, *)) {
       rootView.backgroundColor = [UIColor systemBackgroundColor];
   } else {
-      rootView.backgroundColor = [UIColor whiteColor];
+      rootView.backgroundColor = [UIColor blackColor];
```
Comment (dsznajder) [inline templates/common/ios/{%- project.name %}/AppDelegate.m:54]: Revert it and change backgroundColor to white :)
```suggestion
      rootView.backgroundColor = [UIColor whiteColor];
```

File: templates/common/scripts/postinstall.sh [hunk 6]
```diff
@@ -1,5 +1,7 @@
 #!/bin/bash
 
+git init;
```
Comment (dsznajder) [inline templates/common/scripts/postinstall.sh:3]: What do you think, after the whole process - to create the first commit?

File: templates/navigation/src/navigation/index.tsx [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+import HomeScreen from '~/screens/Home';
+import SettingsScreen from '~/screens/Settings';
+
+import palette from '~/styles/palette';
+
+export type RootStackParamList = {
+  Home: undefined;
+  Settings: undefined;
+};
```
Comment (dsznajder) [inline templates/navigation/src/navigation/index.tsx:15]: Note for future me. Add TS helpers for screen types.

File: templates/navigation/src/screens/Home/Home.tsx [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+import { RootStackParamList } from '~/navigation/index';
+
+const styles = StyleSheet.create({
+  container: {
+    flex: 1,
+    alignItems: 'center',
+    justifyContent: 'center',
+    backgroundColor: palette.primary.background,
+  },
+});
```
Comment (dsznajder) [inline templates/navigation/src/screens/Home/Home.tsx:20]: Normally styles are placed at the end of the file, so please move them. The only reason why navigator had those, was that - I've been preparing it for future theming :)

File: templates/navigation/src/screens/Home/Home.tsx [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+
+type Props = StackScreenProps<RootStackParamList, 'Home'>;
+
+export default function HomeScreen(props: Props) {
+  const { navigation } = props;
+
+  return (
+    <Block style={styles.container}>
+      <Typography>{`DummyProject - Home Page`}</Typography>
```
Comment (dsznajder) [inline templates/navigation/src/screens/Home/Home.tsx:29]: Use text from internationalization.
```suggestion
      <Typography>{t('screens.homeScreen.title')}</Typography>
```

File: templates/navigation/src/screens/Home/Home.tsx [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+    alignItems: 'center',
+    justifyContent: 'center',
+    backgroundColor: palette.primary.background,
+  },
+});
+
+type Props = StackScreenProps<RootStackParamList, 'Home'>;
+
+export default function HomeScreen(props: Props) {
+  const { navigation } = props;
```
Comment (dsznajder) [inline templates/navigation/src/screens/Home/Home.tsx:25]:
```suggestion
const HomeScreen = ({ navigation }: Props) => {
```

Comment (dsznajder) [reply templates/navigation/src/screens/Home/Home.tsx:25]: And at the end of file: `export default HomeScreen;`

File: templates/navigation/src/screens/Home/index.tsx [hunk 11]
```diff
@@ -0,0 +1 @@
+export { default } from './Home';
```
Comment (dsznajder) [inline templates/navigation/src/screens/Home/index.tsx:1]: Move whole `Home.tsx` there.

File: templates/navigation/src/screens/Settings/index.tsx [hunk 12]
```diff
@@ -0,0 +1 @@
+export { default } from './Settings';
```
Comment (dsznajder) [inline templates/navigation/src/screens/Settings/index.tsx:1]: Same as with `Home.tsx`. Move the whole `Settings.tsx` file there.

File: templates/navigation/src/screens/Settings/Settings.tsx [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+    alignItems: 'center',
+    justifyContent: 'center',
+    backgroundColor: palette.primary.background,
+  },
+});
+
+type Props = StackScreenProps<RootStackParamList, 'Settings'>;
+
+export default function SettingsScreen(props: Props) {
+  const { navigation } = props;
```
Comment (dsznajder) [inline templates/navigation/src/screens/Settings/Settings.tsx:25]: 👀

File: templates/navigation/src/screens/Settings/Settings.tsx [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+
+  return (
+    <Block style={styles.container}>
+      <Typography>{`Settings Screen`}</Typography>
```
Comment (dsznajder) [inline templates/navigation/src/screens/Settings/Settings.tsx:29]: 🗒️

**Review by dsznajder — CHANGES_REQUESTED** [review]

File: package.json (hunk 1 above)
Comment (Drakeoon) [reply package.json:39]: I agree 👍🏻  This leaves the `dev` task a little obsolete, so I decided to remove it

File: src/packages.ts (hunk 2 above)
Comment (Drakeoon) [reply src/packages.ts:69]: I initially thought the same, but then I would need to add `modules.navigation` to `if` `ejs` statements in the code. I've decided that `gestureHandler` has it's own responsibility and I don't want to flood it with another module. See example below of a scenario that would happen everywhere:
```
// Before
<% if (modules.gestureHandler) { %>
  
// After
 <% if (modules.gestureHandler || modules.navigation) { %>
 ```

I think this way we don't need to modify conditional statements everywhere, we can just rely on `gestureHandler` doing it's thing, as it is only responsible for making gesture handler work properly

File: templates/common/ios/Podfile (hunk 4 above)
Comment (Drakeoon) [reply templates/common/ios/Podfile:1]: Is this change okay? I think we decided to remove it after our discussions 😄

File: templates/common/ios/{%- project.name %}/AppDelegate.m (hunk 5 above)
Comment (Drakeoon) [reply templates/common/ios/{%- project.name %}/AppDelegate.m:54]: Understood 😄  Is there any reason behind that? For devices below iOS 13 we will always have this weird flicker thing with white background

File: templates/common/scripts/postinstall.sh (hunk 6 above)
Comment (Drakeoon) [reply templates/common/scripts/postinstall.sh:3]: Sounds awesome, I'll add that as this seems to be a quick change

File: templates/common/ios/{%- project.name %}/AppDelegate.m (hunk 5 above)
Comment (Drakeoon) [reply templates/common/ios/{%- project.name %}/AppDelegate.m:54]: Oh, saw your comment above the whole code review notes, now I see 👍🏻

File: templates/common/ios/Podfile (hunk 4 above)
Comment (dsznajder) [reply templates/common/ios/Podfile:1]: Yes, just noting that it caused issues :)

File: package.json (hunk 1 above)
Comment (dsznajder) [reply package.json]: Yeah, that's even better :)

**Review by dsznajder — COMMENTED** [review]
Waiting for rest 👀

**Review by dsznajder — APPROVED** [review]

File: src/packages.ts [hunk 15]
```diff
@@ -48,11 +52,14 @@ const PACKAGES = {
...
         variant: {
           bottomBar: {
-            main: ['@react-navigation/bottom-tabs'],
+            main: ['@react-navigation/stack', '@react-navigation/bottom-tabs'],
           },
           stack: {
             main: ['@react-navigation/stack'],
           },
+          nativeStack: {
+            main: ['@react-navigation/native-stack'],
```
Comment (dsznajder) [inline src/packages.ts:61]:
```suggestion
            main: ['react-native-screens'],
```

File: templates/navigation/src/navigation/index.tsx [hunk 16]
```diff
@@ -5,6 +5,10 @@ import { NavigationContainer } from '@react-navigation/native';
 <% if (extraOptions.modules.navigation.variant === "stack") { %>
 import { createStackNavigator } from '@react-navigation/stack';
 <% } %>
+<% if (extraOptions.modules.navigation.variant === "nativeStack") { %>
+import { enableScreens } from 'react-native-screens';
+import { createNativeStackNavigator } from '@react-navigation/native-stack';
```
Comment (dsznajder) [inline templates/navigation/src/navigation/index.tsx:10]:
```suggestion
import { createNativeStackNavigator } from 'react-native-screens/native-stack';
```

File: src/index.ts [hunk 17]
```diff
@@ -138,9 +193,14 @@ async function create(argv: yargs.Arguments<any>) {
         name: basename,
         package: basename.toLowerCase(),
       },
+      extraOptions,
     } as Options,
   );
 
+  if (options.modules.navigation) {
+    options.modules.gestureHandler = true;
```
Comment (dsznajder) [inline src/index.ts:201]: + `react-native-screens` when `nativeStack` is chosen

```suggestion
    options.modules.gestureHandler = true;
    options.modules.reanimated = true;
```

File: templates/navigation/src/screens/Home/index.tsx [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+  );
+};
+
+export default HomeScreen;
+
+const styles = StyleSheet.create({
+  container: {
+    flex: 1,
+    alignItems: 'center',
+    justifyContent: 'center',
```
Comment (dsznajder) [inline templates/navigation/src/screens/Home/index.tsx:33]:
```suggestion
```

File: templates/navigation/src/screens/Home/index.tsx [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+import Typography from '~/components/Typography';
+import palette from '~/styles/palette';
+import { t } from '~/services/I18n';
+import { RootStackParams } from '~/navigation/types';
+
+type Props = RootStackParams<'Home'>;
+
+const HomeScreen = ({ navigation }: Props) => {
+  return (
+    <Block style={styles.container}>
```
Comment (dsznajder) [inline templates/navigation/src/screens/Home/index.tsx:16]:
```suggestion
    <Block flex={1} alignCenter justifyCenter style={styles.container}>
```

**Review by dsznajder — CHANGES_REQUESTED** [review]
Last polishes and should be ready 👀

File: src/packages.ts (hunk 15 above)
Comment (Drakeoon) [reply src/packages.ts:61]: react-native-screens is always included when we have `navigation` chosen, no matter which variant:

```
navigation: {
      main: [
        '@react-native-community/masked-view',
        '@react-navigation/native',
        'react-native-screens',
        'react-native-safe-area-context',
      ],
    },
```

Comment (dsznajder) [reply src/packages.ts:61]: Okay, so it should be removed

**Comment by Drakeoon** [comment]
@dsznajder I've applied the changes you requested - I also performed manual checks for Android device on each new navigation variant - seems to be all right 👍🏻 I'll check tomorrow on MacOS, but I think this is already in a good shape

**Review by dsznajder — APPROVED** [review]
Thanks!

**Pull request merged**